packages/cli/src/chat/chat-service.ts

```diff
@@ -0,0 +1,346 @@
+		ws.close();
+	});
+
+	setTimeout(() => {
+		if (ws.readyState === WebSocket.OPEN) {
+			ws.close();
+		}
+	}, DRAIN_TIMEOUT);
+}
+
+@Service()
+export class ChatService {
+	private readonly sessions = new Map<string, Session>();
+	private heartbeatIntervalId: NodeJS.Timeout;
+
+	constructor(
+		private readonly executionManager: ChatExecutionManager,
+		private readonly logger: Logger,
+	) {
+		this.heartbeatIntervalId = setInterval(
+			async () => await this.checkHeartbeats(),
+			HEARTBEAT_INTERVAL,
+		);
+	}
+
+	async startSession(req: ChatRequest) {
+		const {
+			ws,
+			query: { sessionId, executionId, isPublic },
+		} = req;
+
+		if (!ws) {
+			throw new UnexpectedError('WebSocket connection is missing');
+		}
+
+		const execution = await this.executionManager.checkExecutionExists(executionId);
+
+		if (!execution) {
+			ws.send(`Execution with id "${executionId}" does not exist`);
+			ws.close(1008);
+			return;
+		}
+
+		ws.isAlive = true;
+
+		const key = `${sessionId}|${executionId}|${isPublic ? 'public' : 'integrated'}`;
+
+		if (this.sessions.has(key)) {
+			this.sessions.get(key)?.connection.terminate();
+			clearInterval(this.sessions.get(key)?.intervalId);
+			this.sessions.delete(key);
+		}
+
+		const onMessage = this.incomingMessageHandler(key);
+		const respondToChat = this.outgoingMessageHandler(key);
+
+		const intervalId = setInterval(async () => await respondToChat(), CHECK_FOR_RESPONSE_INTERVAL);
+
+		ws.once('close', () => {
+			ws.off('message', onMessage);
+			clearInterval(intervalId);
+			this.sessions.delete(key);
+		});
+
+		ws.on('message', onMessage);
+
+		const session: Session = {
+			connection: ws,
+			executionId,
+			sessionId,
+			intervalId,
+			isPublic: isPublic ?? false,
+			isProcessing: false,
+			lastHeartbeat: Date.now(),
+		};
+
+		this.sessions.set(key, session);
+
+		ws.send(N8N_HEARTBEAT);
+	}
+
+	private outgoingMessageHandler(sessionKey: string) {
+		return async () => {
+			let session: Session | undefined;
+			try {
+				session = this.sessions.get(sessionKey);
+
+				if (!session) return;
+				if (session.isProcessing) {
+					return;
+				}
+				session.isProcessing = true;
+
+				const { connection, executionId, sessionId, waitingNodeName, isPublic } = session;
+
+				if (!executionId || !connection) {
+					session.isProcessing = false;
+					return;
+				}
+
+				const execution = await this.getExecutionOrCleanupSession(executionId, sessionKey);
+
+				if (!execution) {
+					session.isProcessing = false;
+					return;
+				}
+
+				if (waitingNodeName) {
+					const lastNode = getLastNodeExecuted(execution);
+
+					if (execution.status === 'waiting' && lastNode?.name !== waitingNodeName) {
+						connection.send(N8N_CONTINUE);
+						session.waitingNodeName = undefined;
+					}
+					session.isProcessing = false;
+					return;
+				}
+
+				if (execution.status === 'waiting') {
+					const message = getMessage(execution);
+
+					if (message !== undefined) {
```

Comment:
What if the node executed but it didn't produce a message? Will the workflow keep waiting forever? Can we detect that and do some sort of error handling there?